src/libsync/networkjobs/resources.cpp

```diff
@@ -0,0 +1,116 @@
+        hash.addData(data.toLatin1());
+        return QString::fromLatin1(hash.result().toHex());
+    }
+
+    QString cacheKey(QNetworkReply *reply)
+    {
+        const QString urlHash = hashMd5(reply->url().toString());
+        const QString eTagHash = hashMd5(reply->header(QNetworkRequest::ETagHeader).toString());
+
+        QString extension = QMimeDatabase().mimeTypeForData(reply).preferredSuffix();
+
+        if (extension.isEmpty()) {
+            extension = QStringLiteral("unknown");
+        }
+
+        return QStringLiteral("%1-%2.%3").arg(urlHash, eTagHash);
+    }
+}
+
+void ResourceJob::finished()
+{
+    qCInfo(lcResources) << "Resource job of" << reply()->request().url() << "finished with status" << replyStatusString();
+
+    if (reply()->error() != QNetworkReply::NoError) {
+        qCWarning(lcResources) << "Network error: " << this << errorString();
+    } else {
+        _cacheKey = cacheKey(reply());
+    }
+
+    // storing the file on disk enables Qt to apply some optimizations, e.g., in QIcon
+    // also, specifically for icons, loading from disk allows easy management of scalable and non-scalable entities
+    if (reply()->size() > 0) {
+        const QString path = _cache->path(_cacheKey);
+
+        qCDebug(lcResources) << "cache file path:" << path;
+
+        // furthermore, we can skip writing the file if the cache key has not changed (i.e., a file exists) and the file has come from the network cache
+        if (QFileInfo::exists(path) && reply()->attribute(QNetworkRequest::SourceIsFromCacheAttribute).toBool()) {
+            qCDebug(lcResources) << "file has come from network cache, skipping writing";
+        } else {
+            QFile cacheFile(path);
+
+            // note: we want to truncate the file if it exists
+            if (!cacheFile.open(QIODevice::WriteOnly)) {
+                qCCritical(lcResources) << "failed to open cache file for writing:" << cacheFile.fileName();
+            } else {
+                if (!cacheFile.write(reply()->readAll())) {
+                    qCCritical(lcResources) << "failed to write to cache file:" << cacheFile.fileName();
+                }
+            }
+        }
+    }
+
+    SimpleNetworkJob::finished();
+}
+
+QIcon ResourceJob::asIcon() const
+{
+    // storing the file on disk enables Qt to apply some optimizations (e.g., caching of rendered pixmaps)
+    Q_ASSERT(!_cacheKey.isEmpty());
+    return QIcon(_cache->path(_cacheKey));
+}
+
+ResourceJob::ResourceJob(const ResourcesCache *cache, const QUrl &rootUrl, const QString &path, QObject *parent)
+    : SimpleNetworkJob(cache->account()->sharedFromThis(), rootUrl, path, "GET", {}, {}, parent)
+    , _cache(cache)
+{
+    setStoreInCache(true);
+}
+
+ResourcesCache::ResourcesCache(const QString &cacheDirectory, Account *account)
+    : QObject(account)
+    , _account(account)
+    , _cacheDirectory(QStringLiteral("%1/tmp.XXXXXX").arg(cacheDirectory))
```

Comment:
The parent dir of the tmp dir must exist

Reply:
`QDir::mkpath` actually could handle this; I wonder why they're not using that. We can fix that by creating the `cacheDirectory` in the caller, but maybe we should just pass the final path to this class and leave creating all dirs to the caller.